addons/l10n_it_edi/models/ir_attachment.py

```diff
@@ -1,20 +1,115 @@
 # -*- coding: utf-8 -*-
 
-from odoo import api, models
+from odoo import _, api, models, tools
 from odoo.addons.l10n_it_edi.tools.remove_signature import remove_signature
+from odoo.exceptions import UserError
 
 from lxml import etree
+from io import BytesIO
+import base64
+import binascii
 import logging
 import re
 
 _logger = logging.getLogger(__name__)
 
+try:
+    from asn1crypto import cms
+except (OSError, ImportError) as err:
+    _logger.debug(err)
+
 FATTURAPA_FILENAME_RE = "[A-Z]{2}[A-Za-z0-9]{2,28}_[A-Za-z0-9]{0,5}.((?i:xml.p7m|xml))"
+re_base64 = re.compile(rb"^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)?$")
+
+
+def is_base64(s):
+    s = s or b""
+    s = s.replace(b"\r", b"").replace(b"\n", b"")
+    return re_base64.match(s)
 
 
 class IrAttachment(models.Model):
     _inherit = 'ir.attachment'
 
+    @api.model
+    def remove_xades_sign(self, xml):
+        # Recovering parser is needed for files where strings like
+        # xmlns:ds="http://www.w3.org/2000/09/xmldsig#&quot;"
+        # are present: even if lxml raises
+        # {XMLSyntaxError}xmlns:ds:
+        # 'http://www.w3.org/2000/09/xmldsig#"' is not a valid URI
+        # such files are accepted by SDI
+        recovering_parser = etree.XMLParser(recover=True, resolve_entities=False)
+        root = etree.XML(xml, parser=recovering_parser)
+        for elem in root.iter("*"):
+            if elem.tag.find("Signature") > -1:
+                elem.getparent().remove(elem)
+                break
+            if any(" " in elem.nsmap[tag] for tag in elem.nsmap):
+                etree.cleanup_namespaces(elem)
+        return etree.tostring(root)
+
+    @api.model
+    def strip_xml_content(self, xml):
+        recovering_parser = etree.XMLParser(recover=True, resolve_entities=False)
+        root = etree.XML(xml, parser=recovering_parser)
+        return etree.tostring(root)
+
+    @staticmethod
+    def extract_cades(data):
+        info = cms.ContentInfo.load(data)
+        return info["content"]["encap_content_info"]["content"].native
+
+    def cleanup_xml(self, xml_string):
+        xml_string = self.remove_xades_sign(xml_string)
+        xml_string = self.strip_xml_content(xml_string)
+        return xml_string
+
+    def get_xml_string(self):
```

Comment:
**Cades:**

We already have our way to remove Cades with PyOpenSSL https://github.com/odoo/odoo/blob/master/addons/l10n_it_edi/tools/remove_signature.py#L17

I also had to add a workaround for invalid Cades signatures (like the ones coming from ex-Enel Servizio Elettrico Nazionale), see: https://github.com/odoo/odoo/pull/137572

Maybe this code is better, I'd be favorable, but let's be careful not to duplicate efforts.

**Xades:**
 
We generally don't need to remove Xades signatures as it's valid XML and we are able to process it, but I guess that we need to remove it for the preview.

**base64:**
We also do the double-base64-decoding in the IAP code when we download it from the Tax Agency channel. From what I understand, this happens when the sender doesn't realize that they are sending base64 and they re-encode the content.

In case it's the client uploading double-encoded base64, I don't think we should accept it as a valid XML file. 

Checking the whole file to see if it's `base64` is kinda slow -- and it slows the main use-case for an exception. We could check just a chunk of it, or go by try-except directly.

Reply:
>I need to know if after the self.raw in the variable have base64 object, because if it isn't base64 I mustn't encode.

If you have an attachment with the `.xml` extension that after decoding it using `raw` it's still `base64` encoded, well then this is an error in itself: we shouldn't allow the upload in the first place.